Gallery view

ShaunYearStrong opened the pull request:
![gallery-load](https://cloud.githubusercontent.com/assets/2104095/15548092/a553d556-229e-11e6-91a5-d4fe7cb2b36b.gif)

ShaunYearStrong commented on kahuna/public/js/components/gu-lazy-gallery-shortcuts/gu-lazy-gallery-shortcuts.js:
```diff
@@ -0,0 +1,40 @@
+                return (event) => {
+                    // Must cancel any scrolling caused by the key
+                    event.preventDefault();
+                    lazyGalleryCtrl[fnName]();
```
This doesn't work, any help would be great.

kenoir replied:
Do you mean that it doesn't prevent the default scrolling event?

ShaunYearStrong replied:
I meant that it didn't call the function. I've since worked out it's because the names of the functions were wrong and Im a moron.

kenoir commented:
@ShaunYearStrong happy to help with Observable stuff - what were you thinking needed improving specifically.

ShaunYearStrong commented:
@kenoir Everything? haha. Nah, I guess I'd just like things to be a bit more reactive. Also I'd like help with the lazy loading logic in terms of reactive programming. Definitely haven't quite got my head around it yet.

paperboyo commented on kahuna/public/js/components/gu-lazy-gallery-shortcuts/gu-lazy-gallery-shortcuts.js:
```diff
@@ -27,13 +27,13 @@ lazyGalleryShortcuts.directive('guLazyGalleryShortcuts',
                     description: 'Go to the previous image',
                     allowIn: ['INPUT'],
-                    callback: invoke('previousImage')
+                    callback: invoke('previousItem')
```
`Item` not `Image` - very prudent: in preparation of Grid support for video ;-) Absolutely innocent ping: @akash1810.

paperboyo commented:
I’ve got so many thoughts on that functionality, maybe I’ll write an email. But then - it would be too long...

Even if *nothing* goes according to my sick plan :smiling_imp: - it’s already a :gem:!

ShaunYearStrong commented on kahuna/public/config.js:
```diff
@@ -95,6 +95,10 @@ System.config({
     "github:jspm/nodelibs-util@0.1.0": {
       "util": "npm:util@0.10.3"
     },
+    "github:tapmodo/Jcrop@0.9.12": {
```
Not sure how this got in here...

akash1810 commented on kahuna/public/stylesheets/main.css:
```diff
@@ -1087,6 +1093,162 @@ textarea.ng-invalid {
 }
 
 /* ==========================================================================
+   Gallery
```
can we extract these styles to their own file?

NickPapacostas commented on kahuna/public/js/components/gu-lazy-gallery/gu-lazy-gallery.js:
```diff
@@ -0,0 +1,144 @@
+    ctrl.init = function({items$, preloadedItems$, currentIndex$}) {
+        const itemsCount$ = items$.map(items => items.length).distinctUntilChanged();
+
+        const buttonCommands$ = new Rx.BehaviorSubject('galleryStart');
```
I think this is fine, but I'm wondering: if you did it the other way (where you assigned all the handlers within the creation block) could you avoid using BehaviorSubject? 

I'm asking because I was just looking at the docs and behavior subject seems to be for "[a value that changes over time](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/subjects/behaviorsubject.md)" which I don't think really applies to the buttonpress stream. 

just something to be aware of

ShaunYearStrong replied:
Hmm true, I'll take a look. When I initialise in the creation block I can add something straight to the stream.

paperboyo commented:
Thinking hard how to get rid of the buttons at the bottom, so that the image can be even bigger (than just by removal of caption area thanks to info panel hook-up). Not related would be getting rid of home/end buttons. I would say it's not likely that they will be used a lot (but it's **great** that home/end keys work!).

Maybe the remaining two previous/next buttons could be moved into (already too crowded, though) bar at top... Hmm. The other solution would be to have them at the sides of the image - but I'm against because it's a long way for a mouse from one to the other then.

There is also a radical option - no visual buttons at all. But with that I would wait for proper fullscreen/everything hidden.

Just thinking out loud...

NickPapacostas commented:
👍  looks good. Agree with @paperboyo but all in due time. 

Nice job ^_^

NickPapacostas commented:
![screen shot 2016-05-26 at 17 26 46](https://cloud.githubusercontent.com/assets/1441582/15582070/0d933198-2367-11e6-91ba-cbb307b6263e.png)


just noticed this if you scroll down

ShaunYearStrong commented:
@NickPapacostas Good spot! This is why we test

akash1810 commented:
couple things I noticed:

- clicking `last` scrolls through all images
- when new images come in, should you be able to scroll left to see them? (not in the scope of this PR)
- its amazing 🎉

ShaunYearStrong commented:
@paperboyo @akash1810 @NickPapacostas - I've fixed the hover area and removed the first/last image logic for now, with a view to moving to a sort of circular gallery i.e. Pressing "Next" on the last image would take you to the first and vice-versa. I think this should be an improvement though, rather than holding up this going out.

NickPapacostas commented on kahuna/public/js/preview/image.js:
```diff
@@ -87,6 +90,41 @@ image.directive('uiPreviewImage', function() {
+                    return Rx.Observable.fromPromise(imgops.getFullScreenUri(image));
+                });
+
+                scope.$watch(() => ctrl.image.data.id, () => {
```
for my own knowledge: is this faster than checking the object?

ShaunYearStrong replied:
It's because this means we're only checking if the actual image has changed, as opposed to the whole object, which also stores things like labels and the like. We don't want to update the visible image when these are added.

NickPapacostas commented:
👍 🉐

akash1810 commented on kahuna/public/js/components/gu-lazy-gallery/gu-lazy-gallery.html:
```diff
@@ -0,0 +1,60 @@
+            </label>
+        </div>
+
+        <div class="gallery-item__preview"
```
not a blocker - would be good to confirm this is needed in this view

akash1810 replied:
maybe add some tracking?

ShaunYearStrong replied:
Will do! The additional div was needed when things were transcluded I'll see if it is now

NickPapacostas replied:
Not sure if you meant Mixpanel tracking @akash1810 but we no longer have a valid MP account, so don't do that @ShaunYearStrong  ^_^

akash1810 replied:
Shoot forgot about that! Maybe we should move to GA at some point?

akash1810 commented on kahuna/public/js/components/gu-lazy-gallery/gu-lazy-gallery.html:
```diff
@@ -0,0 +1,60 @@
+        <button class="gallery__control gallery__control--left"
+                ng:click="galleryCtrl.prevItem()">
+            <gr-icon>navigate_before</gr-icon>
+            <span class="visibly-hidden">Previous</span>
```
is the `span` needed if we're just hiding it anyway?

ShaunYearStrong replied:
For accesibility reasons, yes. If a screen reader user were to tab to this arrow, they'd have no idea what it did. I could always change it to a `title` attribute instead, but this is also pretty standard.

NickPapacostas commented:
commented on an outdated diff >.<  just saw the tracking point. i think this should be removed b.c our mp account is invalid at the moment @ShaunYearStrong 

Sorry for the delay on this.

kenoir commented:
:+1: :+1: :+1:
